### ISSUE TITLE
ACVP testing: Buffer Size Limit Reached when testing HKDF

### ISSUE THREAD

**Issue by abkarcher** [openssl/openssl#19909]
Hello!

In ACVP testing HKDF (SP800-56Cr2), there are times when the fixed info used in HKDF testing can be longer than 2048 bytes. However, the buffer max for fixed info (HKDF_BUFMAX) is a fixed array allocation set to 2048. Thus, ACVP testing cannot pass this implementation. Here is an example ACVP test case:

```
"tgId": 2,
"testType": "AFT",
"kdfConfiguration": {
    "kdfType": "hkdf",
    "l": 2048,
    "saltLen": 512,
    "saltMethod": "default",
    "fixedInfoPattern": "algorithmId||l||uPartyInfo||vPartyInfo",
    "fixedInfoEncoding": "concatenation",
    "hmacAlg": "SHA-1"
}

{
    "tcId": 34,
    "kdfParameter": {
        "kdfType": "hkdf",
        "salt": "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
        "z": "D9B68E36D120291A9DF720D07C932C24E1A5CC0078E704F548B09F9619A621064685FBF4DBF2D00495D770EB5849C9FACCCE0349497312D4AB6A28A124BEF20DE37919468EF9F2E5EE16738BBBB0607152C4550BB4047F60CB89EECDD7BF1B09632C9CB8BB1CBF21D9B74DC9D3DE5E9938006140C8553A7F31D5624D1ACB9D3CE0466A1A4A9871B3041D61EE83B1352ED7099874FCFB4FCD46E6502CD0C43E3AA5D17A080C36A2D1755D3C18238065810232125905420E5A4F7650ED69E6114135846EE9333760E9629EB70A569F4497D3B913E10574BE14BBAD406D20056F2C6893B76717393B151417CDC457CA0890E470F2D8137AD4ECEC86236E4C740130630FD45F2F2908C692275BDA61AA9BDE1C8DE22136D33C3DF2E7902000CE3C70F12E6FF3A9241DC79229AC4FB7C7D47CA7233C89E5F25FD2027D844DF09AF03E7613A1352742025916C71496A53EBB773481BBF00CB38392649735AB7D348FA6BF518E723E09FE6214493D6885C469382902817F41AF900586DF3F26A417A25430995A0F76E479B601FCA09DB90972A3BDDAC22088BF401A610557B8BFFF5EB4E6F672F8682712EC22093E81CAF20ADE98FE2503CB894AA291C23788D19CBF3F7D6A8FC3E80AA2E3DD8EA9D261D2DD998E29547F3FDC1F76BA5A030A0AEF7BCCDCA053A78B9807EC7D65DFEBFF04FC95CD42784C689CE0D1111B5CEFF71F48BC7AA9D75E67AADB0675BD1B1FAD31911649FA2A53ACF575B3FEBA51679F7DA998858C1C911B468FF7952441DE50DF8B6382FC6CA1B8CD49973AF3A4F7A2702BB79B6BFE39951313B69CE321F58CAFA7A3B1F4D8CB34450384C6B2A52FFC2CDBB61ACD8DF7823D585AC7A1E6BC510F49DF22A1237D27468529628176ED6EDB57839F7205D58EC945D571FF11B96E66053C9F130A54AF7A2920A31CC0C812CA59D95B5522574CE8F22CF6D2F751662845234A89B7C8B4B2D8C1CA52BD81F63BA7695078E2B4894FC556DBBE07873CF372619DB703BA33AA513D09ABA7123ECDEBFD5A3445D989E2DA4E43EB7306F0E16D7164FC4EE139029D95526C96639E09414EC5F769A965D75DEC993DBDD140A3AAA00F451B801CB0F4C4DDABB0E7CF738B00BCB3AA9A827DB7807C6B922EC77B41025366D7290AFE026F92EC271D57DA572C6B844D81A6FD06E76E267E1E2086B004C808C1269FF6850A355198886426A6FB89B3B57880090B158C771F8F0837C02A758C43FED6E1DF20FDCD8BB213935F8EB58756E1B126900C0798087F1682D8C003AB1CC4B45C83E094669E111612AD94F73611E1F8C4D59150F85A59C173869BDFA1C5C02856C4C7684FA586B7E10561F2C9F29C1DB762C86B441A7923EC2F39EC55F21B749F5EE9D32F985B8C40D56A717D46F1AD72F1D1BD3960864097F601AA31C78C4B77DF38BAA00F0E891F02CC",
        "l": 2048,
        "algorithmId": "CB5A101AEDD96BB3E23E9254D8198AE2"
    },
    "fixedInfoPartyU": {
        "partyId": "37955F89E76200195B36B45D27432D8C",
        "ephemeralData": "9F11644AFBD66FC2C3E9504B400506C0FAA8DD07095A8D2E6D3D4BF2D302B94B03E341D059692B3AA49A6BF8621231656EB017449C7AEBE1E2BDB32711135D2FE1B35D726B757012EC7A82F53B6FEC20655C0631319BFF6056AD45C97B18B5268FA6684E3397C274E149DF9BDB418E27872F3FC63AECD5490190092EEA525BCB165D36CA977350A523401E07E400F05FDC2F1CD7E4CB84C670FFC2833CE5B5FF51C49CCBA16B57A4072391BEB34649C4134614A052AD90B329D735EF1857C8B4FFD14E2EC404A6F0D06E84747EAD9DA6EFDADA059AB5D09FF261D0475B27138B98106C50EC389EEE4BAE55234C8214F91F91162BE361C56C223A6663F2F19F2592D1F693A7BD370605BE666022F3D959C4C338FB4DA55D8FA4140E243EA4ACB10F781626DBD9BD578C64D040F06980305EA5079FA2AE9DA8F7D26D6415C46AD6FE56C9A32338FECC62DBAAEB4D5562117823E8A97BAB08D81B395FDAA61F3336501E95527AF953474E834AAB87B9E156AD138448FABAF594A9439D76804D50AD8631A42FAB0F24FD6BE363703FB11CB8B0DF27F6CCBA67F44E41619FF7BF2AB2F4895CA3719236D50A500362AF44D50BE39435AAFF4436E3D416C1E2E42F4E6E4513B415BCFF4CE589273B039D3C0B102346CE452E70C5A99E19A3BE77BFFC4554A09FBA54F951B1D9631AC5EDD5FC77D439A2E174A20FE5CE090E0EB12649A4EC28D99AB4E04AC52843E83D5594B2AD2754E4C164E96E2D3B539B9EBF7AF5C1A16F599F8324CCB9FB06CBE3D04B3869A8CE88A6E7EB1446C85CBF809945CCF326B8888DFD5A7BB0FBCCAC7830368AFAD334E48B6E1B7A2AFD56FAFFD5C785A87EFB195B450B9D8C5AF7C26215119DC25B94471FED4AEF4014CE999BE6D5A7BD2EC28ACDF9E68E1A53D2D65D5C8FF996F2E1D47C84B46A09C91341D426F15E30551EF10EA88FA43010CBD35854B85456A2BF8D4506FCDC20D1335C9E169FE6272A2A16B0E59B567F26DB7960D53C3C25B158730C18895B6BCFD6AAFE0ECB0B8A2CA89490DE06AE16CBA5CE6268ABF77C316B69222912CFBEBBF5B693B12125F53DE20BCCB86D911B7B9C1EE8481661856A5C2034C0496C38754CFD702129B3EA8E1679C93162D8519F2A6C954A450BB965D8BD640BB30CB6AB3C2E90BF3145E732692D53CFC65E28C12CE7CC001774A3DDBDA16711E553BC6A48908F8A9A7E87F76CADBFE8E876063FCF4A5F7076FF3BCBB82A3DB93C54E029163B46CD9DF70C8C58430053405C07FB5AA366C57CD2A58F5D2C81A60AE03A6D010045F07261E33510DC431C8F99F83AD1489976C1E1295C2CB92E802C1C689B5F8FEC3996D943F31A27A6A62ED2488CADAEC2737B99FCEE39DCC9B655FD5DB68372EED725A5E893BB1410BDCF6338F66F847D89D3EC0D4D6612DF7F6CFE082AE5760967C8B710"
    },
    "fixedInfoPartyV": {
        "partyId": "5C9C90106A34A4F41A7CDC148A507707",
        "ephemeralData": "CBC1FE44984017D64FF98605793AAE8DB1A4B00D4776C7C5D687CA5AAA9A0355874FFE71FD281120590FC11147B9E1B2228E7AD2AF990EC932FB6CA96171E5100BCEBF5BEDD11C2C775FA6553FFAC043302BCC90AA0E2446AAEB42195488800065000B08FD12DEEE1E25E0D4DFB8B57415C131E2724A1117D7BC6DCBBF3B515833860EF56FD605620D3EDBBC25FC0218B70255A6E0B38673F67BDE0274EC2DF9E106D0E2D06C776A671295F8BD0792B6BEE2B8ED35335F0368DE61E6406D74A33798C3B8B2C1FA6FEF8AD93FB7EF5157611F6FE9906492A5BF83A4CBDC66F441514DE70F5FF4D62ECC4852C8D81E870DAA3D4C35AEC6F11EA822AC88EFC7B84152B25E3FFCFC95AE541EB082833BC76C0C949394A847F4AF8CE40CF4A22A4C7DCAD603A08A93A13ED2F57CA3767E03B4174FB33C4482185653E8D29E3A67A3D1C1260087F0DFBBBFD0D816C09D47985F82475C46FC0DA3CB5DF38046537216E9B61AEA0EDD85FB7495B049E84D754AD32F2510505059ECAC35DE0CEC5350701AC06436E71BC7849297AD9FE9B9D9A969E0D396C38CD64DEC0206D662179321359A03C1F7DF2AA7C5E89C279E46EFB20B40237F75850A1D815B1511C38D6D2FBF4CE6D86BE5E98D5F491CCEAA58A90214C963CF46526D1B726FA6E06C8AF0E2B7477F4849037B6F0F6AD2E2A17D8D283C829BB36F24763E942FDCE754F9F8322CBD4B3780E685B07CD3F3BC91024B4F56FBDA64BC1C9F027E002ECEB4F842022764E5ECCCE42317E758D00F26F43FC3FE90E097FB1F9711AA181BA4FD6C094DE21AB1D1C549DF3792D684542D491B9AA42C3BBD45251DD835AA35AC1BB6ED6CACFE58418FE4EF394E26BD19F8CCBBDFC4375AEBD8B64B3B5C40B5B4822D5D4FEF7C20E929DFD1FF17147391607919EB5EB8DA7806DCE8CDE7822161E4074DC1B71D9211B81A989EB061C7C18332BE7DA199BBCFEF156D818DD803C17E8161DFD5F7C41EF7BB851A88C98D745E76A6EA44A0E245E623B97B3C89CC3D3420683DD525372A1D021E84DFA1F3A891AA0A5A1538E2E6BF36040D21E121153F64BC7B711E1E68D0BC222786A27CEA30DA5D5E9E0CA1F19BD95CF38E08E6402AA4265B8C2383EE08A966B7297EC01C2A931608214BB966BF499FF50D271237734CA546BEA67D47E37FAA5872613A3ACE6A1AA0405A8CCFEBE5010A6A666EB201D91864B98F8644335763AE1BA54CC153095B7BDA14118CB05ED645EBB19E462DC5FDBF228D442AE6F4B551E73D2EC12BECBB09AD9B78FF983BF2FEBA6C39E8A84AEB4A6777AE0E1209175214A9C7A25D90B2249A4791180F6437D09C5734FD47C9FBAEDF8B52DE1156503EB48864CF0B8E55EE6A81B2A4F9CBD5B955BF656348CCDB7879D331B37E1CA64CEB8AE3E51E8C9001BA7CC923417582113ED0FAB7A23CC9986E"
    }
}
```

ephemeralData is optional for both uPartyInfo and vPartyInfo, but when both are provided, and both 1024 bytes long, there is not enough room in the buffer to hold the desired fixedPartyInfo.

There was some discussion in #14649 about dynamically allocating this buffer if the need became evident; I feel that conversation becomes relevant again :)

Thanks!
Andrew

**Comment by paulidale** [comment]
I wonder how our validation passed.  It's entirely possible our lab specified a maximum length for this.

**Comment by abkarcher** [comment]
Having tested recently too, I believe it may have been triggered by recent updates to the NIST demo server for KDA. I have not tried on Prod as I am not currently testing anything specific.

Andrew

**Comment by t8m** [comment]
If it is a bug, it should be fixed on 3.0 as well.

**Comment by slontis** [comment]
I am not sure that having an upper bound is a bug. Its an implementation detail. Regardless of if uses malloc or not I would still prefer some upper bound on this value.
